### ISSUE TITLE
Why GPUCommandBuffer and GPUCommandEncoder cannot be reused?

### ISSUE THREAD

**Issue by microaaron** [gpuweb/gpuweb#4138]
Why once a GPUCommandBuffer object has been submitted, it cannot be used again?
And is it possible to make a copy of the unsubmitted GPUCommandBuffer?
I wish I could put `createCommandEncoder()` and `beginXXXPass()` in the initialization.
Like this:
```
var commandBuffer;

function init(){
  var commandEncoder = device.createCommandEncoder();
  var pass = commandEncoder.beginXXXPass();
  pass.end();
  commandBuffer = commandEncoder.finish();
}

function process(){
  var buffer = commandBuffer.copy(); //Is it possible to make a copy of the unsubmitted GPUCommandBuffer like this?
  device.queue.submit([buffer]);
}
```
I know that GPURenderBundle has similar functionality, but it doesn't work for compute pipelines.
Why not reuse GPUCommandBuffer or GPUCommandEncoder?

**Comment by Kangz** [comment]
This was discussed a very long time ago. The basic reason is that the underlying APIs don't necessarily allow reusing command buffers (Metal, D3D12) or marking command buffers reusable could make them more expensive (Vulkan). So for WebGPU we chose to not allow reusing command buffers. However there was an idea that we could add something like `device.createCommandEncoder({reusable: true})` to allow this use case one day.

**Comment by microaaron** [comment]
> However there was an idea that we could add something like `device.createCommandEncoder({reusable: true})` to allow this use case one day.

Good idea. Thanks for your reply.

**Comment by magcius** [comment]
It's possible you want render bundles which allow pre-recorded commands to be played back on another command buffer. Is there a reason you would like reusable command buffers?

**Comment by kainino0x** [comment]
They said they wanted to use it for compute. I think this is very reasonable. I would really like to have a solution to replaying commands in some way, even if it has no backend equivalent. There's simply no escape from the fact that it's costly to call from JS -> web engine -> gpu process, and that's universal across implementations.

cc: #1003 #286

**Comment by litherum** [comment]
https://developer.apple.com/metal/Metal-Feature-Set-Tables.pdf says that compute ICBs in Metal are available everywhere rendering ICBs are available.

There is an OS requirement, though, where compute ICBs were introduced in macOS 11.

**Comment by kainino0x** [comment]
Both reusable command buffers and compute or generic bundles are implementable even without ICBs (or any API support on any backend) by record-and-replay, so I don't think anything prevents it from being part of core.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/GPU-Web-2023-06-07-08-(Pacific-time)-Meeting-Minutes#intel-why-gpucommandbuffer-and-gpucommandencoder-cannot-be-reused-4138">GPU Web 2023-06-07/08 (Pacific time)</a></summary>

* JS: wonder whether we can have something like RenderBundle for compute shaders. We find we keep re-encoding the same command buffers over and over again for compute use cases. Consumes a lot of CPU time.
* KG: are you concerned about this because of the extra work you're doing? Are you running into perf issues from reencoding these buffers?
* JS: yes, ML models have hundreds of dispatch calls. Have to call them again and again.
* MM: if we added this - how much perf increase would you expect?
* JS: we haven't tested, but we see hundreds of dispatches. Consume significant CPU time.
* KG: concerned more about the CPU time and less about the number of calls. RenderBundles encapsulate thousands of calls - more concerned about that. Caveat - JS calls are expensive in engines. State of the art - what Vulkan says - don't. Just re-encode all the time. Would like to rely on that.
* GT: looking into RenderBundles (not compute). For 10K calls, 2.3 ms -> 1.1 ms (in JS). Optimizing, went from 3 ms -> 1.1 ms. Pretty significant. Had a Chrome issue where passing 10K RenderBundle references in 1 JS array is slow - need to fix that. It's fast in Firefox. A Unity game is doing 3-4K at least. Not sure about ML stuff.
* KN: right now JS calls are a little expensive. No matter how fast they get, 1000 JS calls + 1000 backend calls is more expensive than 1 JS call + 1000 backend calls. At least, costs you power. We don't know how fast they'll be in the long term. Making JS calls and getting them over to the other process takes a lot more work than doing it one time. Think it's worthwhile to have something. It's a straightforward API letting you record/replay. Either something for compute, or something more general than RenderBundles.
* MM: we've got indirect command buffers, and compute indirect cmdbufs, present on all hardware, and we don't see any cost to this. We don't see any obstacles. Should fit right into the API.
* KG: just think it's nonzero eng / impl cost and that we haven't proven to ourselves that this is a problem we need to address. Think it's hypothetical at this point. Technically room to do better, but these are precious investment points and I think we can do better not spending them here.
    * But, if you come to us with a demo and show us a speedup, that would be more compelling.
* JS: understood.
* MM: I found Gregg's numbers compelling.
* KG: would like to see more structure to them before I find them compelling.
* MM: **if Intel runs some measurements is that the next step**?
* KG: yes.
* JS: that's acceptable from our side.

</details>

**Comment by wrightwriter** [comment]
I'm currently facing a situation where i need to re-record the same commands a *lot* (around 20 to 10k) of times, while re-running a couple of render passes in a loop. Targeting low-end mobile devices, so it would be really beneficial to have this in native.     
It's not a "typical" real-time graphics engine, but i can see this becoming an issue in other engines as well.

**Comment by mwyrzykowski** [comment]
> overhead of re-recording the dispatches every frame

Do you mean compute dispatches like dispatchWorkgroups @fridenmf ? Would it help if we had compute bundles too as mentioned above?

**Comment by mwyrzykowski** [comment]
A tiny repro illustrating the overhead would help immensely to justify this too, if it can be provided.

**Comment by fridenmf** [comment]
> Do you mean compute dispatches like dispatchWorkgroups [@fridenmf](https://github.com/fridenmf) ?

Yes.

> Would it help if we had compute bundles too as mentioned above?

Personally I think concepts like RenderBundleEncoder and a future ComputeBundleEncoder only complicates porting an existing Vulkan-like application which already uses some abstraction of re-submittable command buffers. Imagine these command buffers being sent around so that different parts of an engine can fill it with transfer commands, compute commands and render commands on background threads. In the Vulkan backend this is easy to implement as you just call vkCmdCopyBuffer, vkCmdPipelineBarrier, vkCmdDispatch, vkCmdPipelineBarrier, vkCmdBeginRenderPass, vkCmdBindPipeline, ..., vkCmdDrawIndexedIndirect, vkCmdEndRenderPass. It's in the command buffers, and you can simply just submit those over and over on the main thread and you are done.

(*) Now imagine implementing a new WebGPU backend in that same engine. There is no such thing as a transfer bundle, so transfer commands has to be deferred to be re-encoded every frame on the main thread. Then you have the future compute bundles. You can only do compute things in them, so they need their own place in some deferred state that will be executed on the main thread later, but after the transfer commands that came first so you need to bookmark the ordering as well. Then comes render commands which does not fit in compute bundles, they need their own bundles, and they need to be submitted after the compute stuff. Worst case scenario is that you have one transfer command, one compute command and one render command, those three repeated many hundred or thousand times. Then the main thread performance would be just as bad as you have to re-encode them all individually anyways.

As for the RenderBundles, I don't know the whole story, but it seems like it's meant to model something like secondary command buffers in Vulkan which can be re-executed many times in a primary command buffer. But the problem is that RenderBundles only contain render commands. A secondary command buffer in Vulkan can contain transfer commands, compute commands and render commands. Replacing the existing RenderBundle with a more general "CommandBundle" that could be re-executed each frame would be much more usable than yet another specific bundle type like ComputeBundles, cause then you could just prepare big batches of commands a CommandBundle on the background thread, including transfer/compute/render commands, and then re-execute that on the main thread, like a secondary command buffer in Vulkan.

What we would actually want is just to re-submit something that has already been prepared, so that we don't have to spend redundant CPU cycles to say the same thing frame after frame. If the underlaying implementation can already re-submit command buffers, then it sounds to me like exposing that functionality as a optional feature would be the least obscure way of doing that, hence my suggestion in the previous comment.

> A tiny repro illustrating the overhead would help immensely to justify this too, if it can be provided.

I think the example in the (*) paragraph above is enough to realize that you get a lot of redundant work that needs to be done on the main thread as soon as you have "Bundles" specific to render or compute commands. I think it's hard to have a performant main thread for an application with many commands unless re-submission of command buffers is possible or at least having access to a more generalized bundle that can encode transfer, computer and render commands.

**Comment by mwyrzykowski** [comment]
> As for the RenderBundles, I don't know the whole story, but it seems like it's meant to model something like secondary command buffers in Vulkan which can be re-executed many times in a primary command buffer. But the problem is that RenderBundles only contain render commands. A secondary command buffer in Vulkan can contain transfer commands, compute commands and render commands. 

Yes that is my understanding for Vulkan too. In Metal, indirect command buffers can only contain either render or (as of macOS 11) compute commands but not both. I think some UAs still actively support macOS versions <11. So anything other than render commands can't be part of WebGPU core.

> What we would actually want is just to re-submit something that has already been prepared, so that we don't have to spend redundant CPU cycles to say the same thing frame after frame. 

This is not exactly implementable in Metal, which to my understanding is the reason for the current design. However @Kangz 's [suggestion](https://github.com/gpuweb/gpuweb/issues/4138#issuecomment-1564070329) for the `reusable: true` could tell the UA that some identical commands will be submitted multiple times, so on Metal backends the UA could cache to avoid redundant validation and JavaScript -> some other process communication. 

Given the cost of runtime validation observed in some perf traces, this seems like a reasonable benefit to adding a method of reusing command buffers.
